File: src/main/java/eu/mshade/enderframe/world/block/BlockFace.java
package eu.mshade.enderframe.world.block;

import eu.mshade.enderframe.world.Vector;

import java.util.HashMap;
import java.util.Map;

public enum BlockFace {

    UP(1, 0, new Vector(0, 1, 0)),
    DOWN(0, 1, new Vector(0, -1, 0)),
    NORTH(2, 3, new Vector(0, 0, -1)),
    SOUTH(3, 2, new Vector(0, 0, 1)),
    EAST(5, 4, new Vector(1, 0, 0)),
    WEST(4, 5, new Vector(-1, 0, 0)),
    NONE(-1, -1, new Vector(0, 0, 0));

    private int id;
    private int oppositeBlockFace;
    private Vector vector;

    private final static Map<Integer, BlockFace> blockFaceById = new HashMap<>();

    static {
        for (BlockFace blockFace : values()) {
            blockFaceById.put(blockFace.getId(), blockFace);
        }
    }

    BlockFace(int id, int oppositeBlockFace, Vector vector) {
        this.id = id;
        this.oppositeBlockFace = oppositeBlockFace;
        this.vector = vector;
    }

    public int getId() {
        return this.id;
    }

    public BlockFace getOppositeFace(){
        return fromId(this.oppositeBlockFace);
    }

    public Vector getVector() {
        return vector;
    }

    public static BlockFace fromId(int id){
        return blockFaceById.getOrDefault(id, NONE);
    }
}

File: src/main/java/eu/mshade/enderframe/world/block/BlockFace.kt
package eu.mshade.enderframe.world.block

import eu.mshade.enderframe.world.Vector

enum class BlockFace(val id: Int, private val oppositeBlockFace: Int, val vector: Vector) {
    UP(1, 0, Vector(0, 1, 0)),
    DOWN(0, 1, Vector(0, -1, 0)),
    NORTH(2, 3, Vector(0, 0, -1)),
    SOUTH(3, 2, Vector(0, 0, 1)),
    EAST(5, 4, Vector(1, 0, 0)),
    WEST(4, 5, Vector(-1, 0, 0)),
    NONE(-1, -1, Vector(0, 0, 0));

    val oppositeFace: BlockFace
        get() = fromId(oppositeBlockFace)

    fun toAxis(): BlockAxis {
        return toAxis(this)
    }

    companion object {
        private val blockFaceById: MutableMap<Int, BlockFace> = HashMap()

        init {
            for (blockFace in values()) {
                blockFaceById[blockFace.id] = blockFace
            }
        }

        @JvmStatic
        fun fromId(id: Int): BlockFace {
            return blockFaceById.getOrDefault(id, NONE)
        }

        // get block face from direction -180 to 180 degrees
        fun fromDirection(yaw: Float): BlockFace {
            var yaw = yaw
            if (yaw < 0) {
                yaw += 360f
            }
            if (yaw >= 315 || yaw < 45) {
                return SOUTH
            } else if (yaw < 135) {
                return WEST
            } else if (yaw < 225) {
                return NORTH
            } else if (yaw < 315) {
                return EAST
            }
            return NONE
        }

        fun toAxis(blockFace: BlockFace?): BlockAxis {
            return when (blockFace) {
                UP, DOWN -> BlockAxis.Y
                NORTH, SOUTH -> BlockAxis.Z
                EAST, WEST -> BlockAxis.X
                else -> BlockAxis.NONE
            }
        }
    }
}
